### PR TITLE
Post pull requests data to vsm-events-broker /dora

### PR DESCRIPTION
**Pull request by ahmed-ali-55**
Jira: [CID-1905](https://leanix.atlassian.net/browse/CID-1905)

[CID-1905]: https://leanix.atlassian.net/browse/CID-1905?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/feign/FeignDoraProvider.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import net.leanix.vsm.gitlab.broker.shared.Constants.GITLAB_ENTERPRISE_CONNECTOR
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+
+@Component
+class FeignDoraProvider(
+    private val vsmClient: VsmClient
+) : DoraProvider {
+
+    private val logger = LoggerFactory.getLogger(FeignDoraProvider::class.java)
```
Comment (henriqamaral) [inline src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/feign/FeignDoraProvider.kt:17]: Let's try to use the new log library:
private val logger = KotlinLogging.logger {}

File: src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/parser/LanguageParser.kt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package net.leanix.vsm.gitlab.broker.connector.adapter.graphql.parser
+package net.leanix.githubbroker.connector.adapter.graphql.parser
```
Comment (henriqamaral) [inline src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/parser/LanguageParser.kt:1]: I'm nos sure about this package

File: src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/application/DoraService.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                .takeIf { it.isNotEmpty() }
+                ?.forEach { doraProvider.saveDora(it, assignment, repository) }
+                ?: {
+                    logger.warn {
+                        "Repository does not have any valid pull requests for DORA metrics. " +
+                            "Repository: ${repository.name}"
+                    }
+                }
+        }.onFailure {
+            logger.error { "Failed to generate DORA for repository: ${repository.name}: ${it.message}" }
```
Comment (henriqamaral) [inline src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/application/DoraService.kt:35]: Maybe we can send admin logs here

File: src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt [hunk 4]
```diff
@@ -71,6 +77,24 @@ class GitlabGraphqlProvider(private val gitLabOnPremProperties: GitLabOnPremProp
             ?.toRepository()
             ?: throw GraphqlException("No gitlab project found at path: $nameWithNamespace")
 
+    override fun getMergeRequestsForRepository(
+        repository: Repository,
+        periodInDaysInString: String
```
Comment (henriqamaral) [inline src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt:82]: I don't think we need the **String** on the parameter name here.

File: src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt [hunk 5]
```diff
@@ -113,7 +138,25 @@ fun Project.toRepository() = Repository(
...
+    baseRefName = targetBranch,
+    mergedAt = mergedAt.toString(),
+    commits = commits
+        ?.nodes
+        ?.filterNotNull()
+        ?.map {
+            Commit(
+                id = it.id,
+                changeTime = it.authoredDate.toString(),
+                author = Author(it.author?.name ?: "", it.authorEmail ?: "", it.author?.username)
```
Comment (henriqamaral) [inline src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt:156]: I'm not sure if we could accept an author with empty email(I think we use the email to link the metrics with the team).
Can you double-check with team helios how we link the metrics with the team?

Comment (henriqamaral) [reply src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt:156]: But is there any case the emails can be null?
On Bitbucket, it's allowed. So we're filtering them.

File: src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt (hunk 4 above)
Comment (ahmed-ali-55) [reply src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt]: done

File: src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/application/DoraService.kt (hunk 3 above)
Comment (ahmed-ali-55) [reply src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/application/DoraService.kt:35]: done

File: src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt (hunk 5 above)
Comment (ahmed-ali-55) [reply src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt]: done

Comment (ahmed-ali-55) [reply src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt]: the thing is we don't even allow it on **vsm-events-broker**

```kt
data class Commit(
    val id: String,
    val changeTime: String,
    val author: Author,
)

data class Author(
    val name: String,
    val email: String,
    val username: String?,
)
```

File: src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt [hunk 6]
```diff
@@ -8,8 +8,14 @@ import io.github.oshai.kotlinlogging.KotlinLogging
 import kotlinx.coroutines.runBlocking
 import net.leanix.gitlabbroker.connector.adapter.graphql.data.AllGroupsQuery
 import net.leanix.gitlabbroker.connector.adapter.graphql.data.ProjectByPathQuery
+import net.leanix.gitlabbroker.connector.adapter.graphql.data.PullRequestsForProjectQuery
 import net.leanix.gitlabbroker.connector.adapter.graphql.data.allgroupsquery.ProjectConnection
 import net.leanix.gitlabbroker.connector.adapter.graphql.data.projectbypathquery.Project
+import net.leanix.gitlabbroker.connector.adapter.graphql.data.pullrequestsforprojectquery.MergeRequest
+import net.leanix.vsm.githubbroker.connector.domain.Author
```
Comment (henriqamaral) [inline src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt:15]: I haven't noticed before.
package wrong here

Comment (ahmed-ali-55) [reply src/main/kotlin/net/leanix/vsm/gitlab/broker/connector/adapter/graphql/GitlabGraphqlProvider.kt]: Done. thanks

**Review by henriqamaral — APPROVED** [review]

**Pull request merged**